### PR TITLE
Add username check to doctor

### PR DESCRIPTION
**Pull request by Gusted**
- Add a new breaking change detector to Gitea's doctor, which checks if all users still have a valid username according to Gitea. Given from time-to-time we need to make changes, either due to new routes or due to security, it's for a instance's admin to check if all users still have a valid username.

User with invalid username:
image

All users have a valid username:
image

### CODE REVIEW

**Review by wxiaoguang — APPROVED** [review]

File: modules/doctor/breaking.go [hunk 1]
```diff
@@ -66,4 +89,11 @@ func init() {
 		Run:       checkUserEmail,
 		Priority:  9,
 	})
+	Register(&Check{
+		Title:     "Check if users has an valid username",
```
Comment (jpraet) [inline modules/doctor/breaking.go]:
```suggestion
		Title:     "Check if users have a valid username",
```

File: modules/doctor/breaking.go [hunk 2]
```diff
@@ -58,6 +58,29 @@ func checkUserEmail(ctx context.Context, logger log.Logger, _ bool) error {
...
 
+// From time to time Gitea makes changes to the reserved usernames and which symbols
+// are allowed for various reasons. This check helps with detecting users that are according
+// to Gitea don't have a valid username.
+func checkUserName(ctx context.Context, logger log.Logger, _ bool) error {
+	var invalidUserCount int64
+	if err := iterateUserAccounts(ctx, func(u *user.User) error {
+		if err := user.IsUsableUsername(u.Name); err != nil {
+			invalidUserCount++
+			logger.Warn("User[id=%d] have not a valid username: %v", u.ID, err)
```
Comment (jpraet) [inline modules/doctor/breaking.go]:
```suggestion
			logger.Warn("User[id=%d] does not have a valid username: %v", u.ID, err)
```

**Review by jolheiser — APPROVED** [review]

File: modules/doctor/breaking.go [hunk 3]
```diff
@@ -66,4 +89,11 @@ func init() {
 		Run:       checkUserEmail,
 		Priority:  9,
 	})
+	Register(&Check{
+		Title:     "Check if users have a valid username",
+		Name:      "check-user-name",
```
Comment (delvh) [inline modules/doctor/breaking.go]: Given that you don't check for a single user but for all instead, what about using
```suggestion
		Name:      "check-user-names",
```
instead?

File: modules/doctor/breaking.go [hunk 4]
```diff
@@ -58,6 +58,29 @@ func checkUserEmail(ctx context.Context, logger log.Logger, _ bool) error {
 	return nil
 }
 
+// From time to time Gitea makes changes to the reserved usernames and which symbols
+// are allowed for various reasons. This check helps with detecting users that are according
+// to Gitea don't have a valid username.
```
Comment (delvh) [inline modules/doctor/breaking.go]:
```suggestion
// are allowed for various reasons. This check helps with detecting users that, according
// to our reserved names, don't have a valid username.
```

**Review by delvh — APPROVED** [review]

**Comment by jolheiser** [comment]
This command is mentioned in our 1.17 release notes, but is only in 1.18 currently. 
Does this need a backport?

**Pull request merged**